<div class="alert alert-info">
Привет, Дмитрий! Меня зовут Светлана и я буду проверять твой проект. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе. Предлагаю общаться на «ты». Но если это не удобно - дай знать, и мы перейдем на «вы».

<div class="alert alert-success">
<b>👍 Успех:</b> Зелёным цветом отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах.
</div>
<div class="alert alert-warning">
<b>🤔 Рекомендация:</b> Жёлтым цветом выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).
</div>
<div class="alert alert-danger">
<b>😔 Необходимо исправить:</b> Красным цветом выделены комментарии, без исправления которых, я не смогу принять проект :(
</div>
<div class="alert alert-info">
<b>👂 Совет:</b> Какие-то дополнительные материалы
</div>
Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом.
Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Например, вот так.</div>
Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Всё это поможет выполнить повторную проверку твоего проекта быстрее.
 </div>

 <div class="alert alert-danger">
    <s><b>😔 Необходимо исправить:</b> В проект нужно добавить описательную часть</s>
 </div>

<div class="alert alert-info"> <b>🎓  </b>исправил</div>

**анализ книжного магазина**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.


Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель исследования** - изучить данные о книгах, авторах, издательствах, рейтингах и обзорах и ответить на поставленные вопросы 

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [2]:
# список таблиц
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']

#просмотр типов данных
for table in tables:
    query = f"SELECT * FROM {table} LIMIT 0"
    df = pd.io.sql.read_sql(sql=text(query), con = con)
    print(f"Columns in {table} table:")
    display(df.dtypes)


Columns in books table:


book_id             object
author_id           object
title               object
num_pages           object
publication_date    object
publisher_id        object
dtype: object

Columns in authors table:


author_id    object
author       object
dtype: object

Columns in publishers table:


publisher_id    object
publisher       object
dtype: object

Columns in ratings table:


rating_id    object
book_id      object
username     object
rating       object
dtype: object

Columns in reviews table:


review_id    object
book_id      object
username     object
text         object
dtype: object

In [3]:

for table in tables:
    query = f"SELECT * FROM {table}"
    df = pd.io.sql.read_sql(sql=text(query), con = con)

    print(f"Data in {table} table:")
    display(df.head())


Data in books table:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Data in authors table:


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Data in publishers table:


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Data in ratings table:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Data in reviews table:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

 <div class="alert alert-success">
 <b>👍 Успех:</b> Импортированы все нужные библиотеки, создано подключение к БД
 </div>

 <div class="alert alert-danger">
    <s> <b>😔 Необходимо исправить:</b> Перед выполнением заданий нужно ознакомиться с таблицами БД</s>
 </div>

<div class="alert alert-info"> <b>🎓  </b>исправил</div>

In [4]:
# создаем функцию для выполнения SQL-запросов
def execute_query(query):
    return pd.io.sql.read_sql(sql=text(query), con=engine.connect())

# Посчитайте, сколько книг вышло после 1 января 2000 года
query = '''SELECT COUNT(*) FROM books WHERE publication_date > '2000-01-01';'''
display(execute_query(query))

count
0    819

после 1 января 2000 года было впушенно 819 книг

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

Сначала код не работал и я прописал функцию в каждом блоке кода, но потом без этого заработало, если что можно раскоментировать 

In [5]:
# колличество и средняя оценка 
# def execute_query(query):
#     return pd.io.sql.read_sql(sql=text(query), con=engine.connect())

query = '''
SELECT 
    books.book_id,
    books.title, 
    COUNT(DISTINCT review_id) as review_count, 
    AVG(rating) as avg_rating
FROM 
    books 
LEFT JOIN reviews ON books.book_id = reviews.book_id
LEFT JOIN ratings ON books.book_id = ratings.book_id
GROUP BY 
    books.book_id,
    books.title
'''

display(execute_query(query))

book_id                                              title  review_count  \
0          1                                       'Salem's Lot             2   
1          2                 1 000 Places to See Before You Die             1   
2          3  13 Little Blue Envelopes (Little Blue Envelope...             3   
3          4  1491: New Revelations of the Americas Before C...             2   
4          5                                               1776             4   
..       ...                                                ...           ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)             3   
996      997                        Xenocide (Ender's Saga  #3)             3   
997      998                                    Year of Wonders             4   
998      999                        You Suck (A Love Story  #2)             2   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...             4   

     avg_rating  
0      3.666667  
1      2.500000  
2      4.666667  
3      4.500000  
4      4.000000  
..          ...  
995    3.666667  
996    3.400000  
997    3.200000  
998    4.500000  
999    3.833333  

[1000 rows x 4 columns]

 <div class="alert alert-danger">
 <s><b>😔 Необходимо исправить:</b> Название книги не уникальный признак, только по нему не нужно группировать, здесь нужно добаить группировку еще и по book_id</s>
 </div>

<div class="alert alert-info"> <b>🎓  :</b> исправил</div>

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

таблица со средними оценками книн 

In [6]:
# def execute_query(query):
#     return pd.io.sql.read_sql(sql=text(query), con=engine.connect())


# Определите издательство, которое выпустило наибольшее число книг толще 50 страниц
query = '''
SELECT 
    publishers.publisher, 
    COUNT(books.book_id) as book_count
FROM 
    books 
INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE 
    books.num_pages > 50
GROUP BY 
    publishers.publisher
ORDER BY 
    book_count DESC
LIMIT 1
'''
display(execute_query(query))


publisher  book_count
0  Penguin Books          42

издательство Penguin Books выпостило 42 киги более 50 страниц 

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

In [7]:
# def execute_query(query):
#     return pd.io.sql.read_sql(sql=text(query), con=engine.connect())

# Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками
query = '''
SELECT 
    authors.author, 
    AVG(avg_book_rating) as avg_author_rating
FROM 
    (SELECT 
        books.author_id, 
        AVG(ratings.rating) as avg_book_rating
    FROM 
        books 
    INNER JOIN ratings ON books.book_id = ratings.book_id
    GROUP BY 
        books.book_id
    HAVING 
        COUNT(ratings.rating) >= 50) as subquery
INNER JOIN authors ON subquery.author_id = authors.author_id
GROUP BY 
    authors.author
ORDER BY 
    avg_author_rating DESC
LIMIT 1
'''
display(execute_query(query))



author  avg_author_rating
0  J.K. Rowling/Mary GrandPré           4.283844

автор с самой высокой среденей оценкой 4.28 J.K. Rowling/Mary GrandPré

 <div class="alert alert-danger">
<s> <b>😔 Необходимо исправить:</b> Получилось не совсем верно, фильтр HAVING призван отфильтровать книги с оценками 50 и более, а данные ты группируешь по автору. В итоге фильтр фильтрует не книги, а авторов, и в такой группировке почти у каждого автора будет 50 и более оценок, поэтому рейтинг будет рассчитан неверно - в него попадут лишние книги, которые не проходят по условию с количеством оценок</s>
 </div>

<div class="alert alert-info"> <b>🎓 </b> исправил</div>

 <div class="alert alert-danger">
<s> <b>😔 Необходимо исправить V2:</b> Почти, но не хватает еще одного шага. Сейчас ты нашел список автор-книга, теперь нужно еще раз сгруппировать по автору и найти среднюю оценку от средней по книгам автора</s>
 </div>

<div class="alert alert-info"> <b>🎓 </b> исправил</div>

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

In [8]:
# def execute_query(query):
#     return pd.io.sql.read_sql(sql=text(query), con=engine.connect())
# Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок
query = '''
SELECT 
    AVG(review_count) as avg_review_count
FROM 
    (SELECT 
        username, 
        COUNT(review_id) as review_count
    FROM 
        reviews 
    WHERE 
        username IN (SELECT 
                        username 
                     FROM 
                        ratings 
                     GROUP BY 
                        username 
                     HAVING 
                        COUNT(rating_id) > 48)
    GROUP BY 
        username) as subquery
'''
display(execute_query(query))

avg_review_count
0              24.0

среднее количество обзоров от пользователей, которые поставили больше 48 оценок 24

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

In [9]:

#def execute_query(query):
    #return pd.io.sql.read_sql(sql=text(query), con=engine.connect())
# подсчет количества строк в каждой таблице
for table in tables:
    query = f'SELECT COUNT(*) FROM {table};'
    result = execute_query(query)
    print(f'Количество строк в таблице {table}: {result.values}')

Количество строк в таблице books: [[1000]]
Количество строк в таблице authors: [[636]]
Количество строк в таблице publishers: [[340]]
Количество строк в таблице ratings: [[6456]]
Количество строк в таблице reviews: [[2793]]


**Вывод**


С 1 января 2000 года было выпущено 819 книг. Издательство Penguin Books выпустило 42 книги, каждая из которых содержит более 50 страниц. Автор с самой высокой средней оценкой 4.28 - это Rowling/Mary GrandPré Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, составляет 24

колличество строк 
- Количество строк в таблице книг: **1000**
- Количество строк в таблице авторов: **636**
- Количество строк в таблице издателей: **340**
- Количество строк в таблице рейтингов: **6456**
- Количество строк в таблице обзоров: **2793**

<div class="alert alert-success">
<b>👍 Успех:</b> Хорошая работа, поздравляю с окончанием курса! Рекомендую еще пройти бесплатный курс от практикума по <a href='https://practicum.yandex.ru/ycloud-datalens/'> DataLens</a>
   Дополнительную информацию по SQL  можно посмотреть и потренироваться <a href='https://sql-ex.ru'> здесь</a>, <a href='https://pgexercises.com'> здесь</a>, <a href='https://sql-academy.org'> здесь</a>, <a href='https://stepik.org/course/63054/info'> здесь</a>, <a href='https://Itresume.ru'> здесь</a>
</div>